README
Games-Bettor

This module and the associated modules under Bettor/* represent bettors using different types of betting
systems.  Each can be passed various pieces of information that help them decide what their wager on each
trial should be.  

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install


SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the perldoc command.

    perldoc Games::Bettor

You can also look for information at:

    Search CPAN
        http://search.cpan.org/dist/Games-Bettor

    CPAN Request Tracker:
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Games-Bettor

    AnnoCPAN, annotated CPAN documentation:
        http://annocpan.org/dist/Games-Bettor

    CPAN Ratings:
        http://cpanratings.perl.org/d/Games-Bettor

COPYRIGHT AND LICENCE

Copyright (C) 2006 Troy Denkinger

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
Revision history for Games-Bettor

1.00    11/12/2006
        Initial release



scripts/example.pl
#!/usr/bin/perl

use warnings;
use strict;
use lib('./lib');
use Data::Dumper;

use constant TRIALS => 1_000;
use constant BIAS   => .10;
use constant ODDS   => 1;

use Games::FlipCoin;
use Games::Bettor;
use Games::Bettor::Consistent;
use Games::Bettor::Martingale;
use Games::Bettor::Kelly;
use Games::Bettor::Reckless;

my $martin = Games::Bettor->new( name => 'Martin', 
                                 bankroll => 1000, 
                                 method => Games::Bettor::Martingale->new( amount => 1 ),
                               );

my $connie = Games::Bettor->new( name => 'Connie', 
                                 bankroll => 1000, 
                                 method => Games::Bettor::Consistent->new( amount => 1 ),
                               );
                               
my $kelly = Games::Bettor->new( name => 'Kelly',
                                bankroll => 1000,
                                method => Games::Bettor::Kelly->new( winner => 50 + BIAS * 100, odds => ODDS ),
                              );
                              
my $becky = Games::Bettor->new( name => 'Becky',
                                bankroll => 1000,
                                method => Games::Bettor::Reckless->new(),
                              );                               
                                                             
                               
my $max_trials = TRIALS;
my $win_count = 0;
for( my $trials = 1; $trials <= $max_trials; $trials++ ){

  my $flip = flip( bias => BIAS );
  
  $win_count++ if $flip == 1;

  $martin->log_result( result => $flip );  
  $connie->log_result( result => $flip );  
  $kelly->log_result( result => $flip );
  $becky->log_result( result => $flip );  

}

foreach( $martin, $connie, $kelly, $becky ){
  my $status = $_->report();
  print $status->{'name'}, " (", $status->{'betting_method'}, ")\n";
  print "Starting: ", $status->{'bankroll_start'}, "\n";
  print "Current: ", $status->{'bankroll_current'}, "\n";
#  print Dumper $status->{'history'}, "\n" if $status->{'name'} eq 'Becky';
  print "\n";
  
}

lib/Games/Bettor/Kelly.pm
package Games::Bettor::Kelly;

use warnings;
use strict;
use Carp;
use Data::Dumper;

sub new{
	my ( $class, %args ) =  @_;
		
	bless{
	      name       => 'Kelly Criterion',
	      win_chance => $args{'winner'},
	      odds       => $args{'odds'},
	}, $class;
}

sub name{
  my ( $self ) = @_;
  return $self->{'name'};
}

sub get_bet{
  my ( $self, %args ) = @_;

  # percent_of_br_to_bet = (odds received * probability of winning - probability of losing )/odds received
  # bet = bankroll * percent_of_br_to_bet
  my $percent = ( ( $self->{'odds'} * $self->{'win_chance'} ) - ( 100 - $self->{'win_chance'} ) ) / $self->{'odds'} / 100;

  my $bet = $percent * $args{'bankroll'};


  return $bet;
}

=head1 NAME

Games::Bettor::Kelly - A class that holds the methods for a kelly formula bettor

=head1 VERSION

Version 1.0

=cut

our $VERSION = '1.00';

=head1 SYNOPSIS


    use Games::Bettor::Kelly;

    $bettor = Games::Bettor::Kelly->new( winner => 75, odds => 2);

=head1 DESCRIPTION

This method represents a bettor who always bets using the Kelly Criterion system.  Basically, the Kelly bettor
bets only a portion of his bankroll.  This criterion is supposed to balance risk of ruin with the best rate of
return.  Since only a portion of the bankroll is bet, the bettor can never go broke.

=head1 METHODS

=head2 new

  $method = Games::Bettor::Kelly->new( winner => $winning_percentage, odds => $odds );
  
  winner: The percentage of the time you expect to win expressed as a percentage, ie 65.
  odds: The left side of an odds expression.  5:1 would be expressed as 5 and even money odds would just be 1:1 or 1.

=head2 name

  $method_name = $method->name();
  
  Returns the name of the betting method.

=head2 get_bet

  $amount = $method->get_bet( bankroll => $bankroll );
  
  Returns the amount to bet for this trial.  The bankroll argument is the amount of the bankroll currently for 
  the Bettor object using this method.

=head1 AUTHOR

Troy Denkinger, C<< <tdenkinger at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-games-bettor at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Games-Bettor>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Games::Bettor::Kelly

You can also look for information at:

=over 4

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Games-Bettor>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Games-Bettor>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Games-Bettor>

=item * Search CPAN

L<http://search.cpan.org/dist/Games-Bettor>

=back

=head1 ACKNOWLEDGEMENTS

=head1 COPYRIGHT & LICENSE

Copyright 2006 Troy Denkinger, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1; # End of Games::Bettor::Kelly


lib/Games/Bettor/Reckless.pm
package Games::Bettor::Reckless;

use warnings;
use strict;
use Carp;

sub new{
	my ( $class, %args ) =  @_;
		
	bless{
	      name    => 'Reckless',
	}, $class;
}

sub name{
  my ( $self ) = @_;
  return $self->{'name'};
}

sub get_bet{
  my ( $self, %args ) = @_;
  
  my $bet;
  $bet = $args{'bankroll'};
  return $bet;
}

=head1 NAME

Games::Bettor::Reckless - A class that holds the methods for a reckless bettor

=head1 VERSION

Version 1.0

=cut

our $VERSION = '1.00';

=head1 SYNOPSIS


    use Games::Bettor::Reckless;

    $bettor = Games::Bettor::Reckless->new();

=head1 DESCRIPTION

This method represents a bettor who always bets his entire bankroll on each trial.  This is a bettor who is 
almost always guaranteed to go broke.

=head1 METHODS

=head2 new

  $method = Games::Bettor::Martingale->new();

=head2 name

  $method_name = $method->name();
  
  Returns the name of the betting method.

=head2 get_bet

  $amount = $method->get_bet( bankroll => $bankroll );
  
  Returns the amount to bet for this trial.  The bankroll argument is the amount of the bankroll currently for 
  the Bettor object using this method.


=head1 AUTHOR

Troy Denkinger, C<< <tdenkinger at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-games-bettor at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Games-Bettor>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Games::Bettor::Reckless

You can also look for information at:

=over 4

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Games-Bettor>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Games-Bettor>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Games-Bettor>

=item * Search CPAN

L<http://search.cpan.org/dist/Games-Bettor>

=back

=head1 ACKNOWLEDGEMENTS

=head1 COPYRIGHT & LICENSE

Copyright 2006 Troy Denkinger, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1; # End of Games::Bettor::Reckless


lib/Games/Bettor/Consistent.pm
package Games::Bettor::Consistent;

use warnings;
use strict;
use Carp;

sub new{
	my ( $class, %args ) =  @_;
		
  $args{'percent'} = 0 unless exists $args{'percent'} && defined $args{'percent'};
  
	bless{
	      name    => 'Consistent',
	      amount  => $args{'amount'}        || 0,
	      percent => $args{'percent'} * .01,
	}, $class;
}

sub name{
  my ( $self ) = @_;
  return $self->{'name'};
}

sub get_bet{
  my ( $self, %args ) = @_;
  
  my $bet;
  if(     $self->{'amount'}  ){ $bet = $self->{'amount'};
  }elsif( $self->{'percent'} ){ $bet = $args{'bankroll'} * $self->{'percent'};
  }else{                        $bet = 0;
  }

  return $bet;
}

=head1 NAME

Games::Bettor::Consistent - A class that holds the methods for a consistent bettor

=head1 VERSION

Version 1.0

=cut

our $VERSION = '1.00';

=head1 SYNOPSIS


    use Games::Bettor::Consistent;

    $bettor = Games::Bettor::Consistent->new( amount => 10 ); # always bets 10 units
    $bettor = Games::Bettor::Consistent->new( percent => 10 );  # always bets 10% of the bankroll

=head1 DESCRIPTION

This method represents a bettor who always bets the same amount on each trial.  The amount bet can be a flat 
amount of a percentage of the current bankroll, but it's always the same amount or percent.

=head1 METHODS

=head2 new

  $method = Games::Bettor::Consistent->new( amount => 1 );

=head2 name

  $method_name = $method->name();
  
  Returns the name of the betting method.

=head2 get_bet

  $amount = $method->get_bet( bankroll => $bankroll );
  
  Returns the amount to bet for this trial.  The bankroll argument is the amount of the bankroll currently for 
  the Bettor object using this method.

=head1 AUTHOR

Troy Denkinger, C<< <tdenkinger at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-games-bettor at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Games-Bettor>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Games::Bettor::Consistent

You can also look for information at:

=over 4

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Games-Bettor>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Games-Bettor>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Games-Bettor>

=item * Search CPAN

L<http://search.cpan.org/dist/Games-Bettor>

=back

=head1 ACKNOWLEDGEMENTS

=head1 COPYRIGHT & LICENSE

Copyright 2006 Troy Denkinger, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1; # End of Games::Bettor::Consistent


lib/Games/Bettor/Martingale.pm
package Games::Bettor::Martingale;

use warnings;
use strict;
use Carp;
use Data::Dumper;

sub new{
	my ( $class, %args ) =  @_;
		
  $args{'percent'} = 0 unless exists $args{'percent'} && defined $args{'percent'};
  
	bless{
	      name      => 'Martingale',
	      amount    => $args{'amount'},
	      percent   => $args{'percent'} * .01,
	      prior_bet => undef
	}, $class;
}

sub name{
  my ( $self ) = @_;
  return $self->{'name'};
}

sub get_bet{
  my ( $self, %args ) = @_;

  my $bet;

  # if there's no prior bet, this is the first bet, so return the base bet
  # otherwise, check the prior result
  #   if the prior result was a win, use the base bet
  #   otherwise, double the size of the last bet and return that
  if( ! defined $self->{'prior_bet'} ){
      $bet = _get_base_bet( $self, $args{'bankroll'} );    
  }else{    
    if( $args{'history'}[-1]{'result'} == 1 ){
      $bet = _get_base_bet( $self, $args{'bankroll'} );
    }else{
      $bet = $self->{'prior_bet'} * 2;
    }
  }

  $self->{'prior_bet'} = $bet;
  
  return $bet;
}

sub _get_base_bet{
  my ( $self, $bankroll ) = @_;

  my $bet;
  
  if(     $self->{'amount'}  ){ $bet = $self->{'amount'};
  }elsif( $self->{'percent'} ){ $bet = $bankroll * $self->{'percent'};
  }else{                        $bet = 0;
  }    
  
  return $bet;
}

=head1 NAME

Games::Bettor::Martingale - A class that holds the methods for a martingale bettor

=head1 VERSION

Version 1.0

=cut

our $VERSION = '1.00';

=head1 SYNOPSIS


    use Games::Bettor::Martingale;

    $bettor = Games::Bettor::Martingale->new( amount => 1 );
    $bettor = Games::Bettor::Martingale->new( percent => 10 );

=head1 DESCRIPTION

This method represents a bettor who always bets using the Martingale system.  This bettor will always bet the
same amount on the first bet and on any subsequent bet where he has won the previous trial.  If he has lost
the previous trial, he will double his last bet until he wins.  

=head1 METHODS

=head2 new

  $method = Games::Bettor::Martingale->new( amount => 1 );
  $method = Games::Bettor::Martingale->new( percent => 10 );
  
  amount: The amount of bet on each round in absolute betting units.
  percent: The amount fo bet on each round in percentage of bankroll.

=head2 name

  $method_name = $method->name();
  
  Returns the name of the betting method.

=head2 get_bet

  $amount = $method->get_bet( bankroll => $bankroll );
  
  Returns the amount to bet for this trial.  The bankroll argument is the amount of the bankroll currently for 
  the Bettor object using this method.


=head1 AUTHOR

Troy Denkinger, C<< <tdenkinger at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-games-bettor at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Games-Bettor>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Games::Bettor::Martingale

You can also look for information at:

=over 4

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Games-Bettor>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Games-Bettor>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Games-Bettor>

=item * Search CPAN

L<http://search.cpan.org/dist/Games-Bettor>

=back

=head1 ACKNOWLEDGEMENTS

=head1 COPYRIGHT & LICENSE

Copyright 2006 Troy Denkinger, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1; # End of Games::Bettor::Martingale


lib/Games/Bettor.pm
package Games::Bettor;

use warnings;
use strict;
use Carp;
use Data::Dumper;

sub new{
	my ( $class, %args ) =  @_;
		
	bless{
		    name             => $args{'name'},
		    start_bankroll   => $args{'bankroll'},
		    current_bankroll => $args{'bankroll'},
		    method           => $args{'method'},
		    history          => _initialize_history( $args{'history'} ),
	}, $class;
}

sub name{
  my ( $self ) = @_;
  return $self->{'name'};
}

sub log_result{
  my ( $self, %args ) = @_;
  
  my $pre_bankroll = $self->{'current_bankroll'};
  
  my $bet = _get_bet( $self );
  return if $bet == 0; # Our bankroll is depleted
  
  if(     $args{'result'} == 1 ){  $self->{'current_bankroll'} += $bet;
  }elsif( $args{'result'} == 0 ){  $self->{'current_bankroll'} -= $bet;
  }else{
    # Do nothing, we have a bad result
  }
  
  my %history_entry = ( bet => $bet, result => $args{'result'}, pre_bankroll => $pre_bankroll, post_bankroll => $self->{'current_bankroll'} );
  push @{$self->{'history'}}, \%history_entry;
}

sub report{
  my ( $self ) = @_;
  
  my %status = ( name             => $self->{'name'},
                 bankroll_start   => $self->{'start_bankroll'},
                 bankroll_current => $self->{'current_bankroll'},
                 betting_method   => $self->{'method'}->name(),
                 history          => $self->{'history'},
               );

  return \%status;  
}

sub _get_bet{
  my ( $self, %args ) = @_;  
  my $bet = $self->{'method'}->get_bet( bankroll => $self->{'current_bankroll'},
                                        history  => $self->{'history'},
                                      );
  return $bet;
}

sub _initialize_history{
  my ( $prior_history ) = @_;
  my @history;
  return \@history;
}

=head1 NAME

Games::Bettor - A class representing a bettor

=head1 VERSION

Version 1.0

=cut

our $VERSION = '1.00';

=head1 SYNOPSIS

This class represents an individual bettor with a name, bankroll, betting method and history.

Perhaps a little code snippet.

    use Games::Bettor;

    my $bettor = Games::Bettor->new( name => 'Connie', 
                                     bankroll => 1000, 
                                     method => Games::Bettor::Consistent->new( amount => 1 )
                                   );

=head1 METHODS

=head2 new

  $bettor = Games::Bettor->new( name     => $string,
                                bankroll => $number,
                                method   => $ref_to_method
                              );
  
  Returns a new Bettor object.
  
  name: A human readable designation for this object.
  bankroll: The starting bankroll for this bettor.
  method: a reference to the betting method to use.

=head2 name

  my $name= $bettor->name();
  
  Returns the name associated with the Bettor object;

=head2 log_result

  $bettor->log_result( result => [0|1] );
  
  Alerts the Bettor object to the outcome of a paricular trial.

=head2 report

  my $bettor_data = $bettor->report();
    
  Returns a hash reference containing data on the run for a particular Bettor object.  The data points
  returned are name, starting bankroll, ending bankroll and the trial history.


=head1 AUTHOR

Troy Denkinger, C<< <tdenkinger at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-games-bettor at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Games-Bettor>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Games::Bettor

You can also look for information at:

=over 4

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Games-Bettor>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Games-Bettor>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Games-Bettor>

=item * Search CPAN

L<http://search.cpan.org/dist/Games-Bettor>

=back

=head1 COPYRIGHT & LICENSE

Copyright 2006 Troy Denkinger, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1; # End of Games::Bettor
